libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt

```diff
@@ -0,0 +1,296 @@
+        override fun onSuccess(result: LocationEngineResult?) {
+            ifNonNull(result, result?.lastLocation) { _, lastLocation ->
+                channelLocationBuffer.offer(LocationBufferControl(LocationBufferCommands.BUFFER_ADD, lastLocation))
+            }
+        }
+
+        override fun onFailure(exception: Exception) {
+            MapboxLogger.i(TAG, Message("location services exception: $exception"))
+        }
+    }
+
+    /**
+     * Obtains a user agent string based on where this code is being called from
+     */
+    private fun obtainUserAgent(options: MapboxNavigationOptions): String? {
+        return if (options.isFromNavigationUi) {
+            MAPBOX_NAVIGATION_UI_USER_AGENT_BASE + BuildConfig.MAPBOX_NAVIGATION_VERSION_NAME
+        } else {
+            MAPBOX_NAVIGATION_USER_AGENT_BASE + BuildConfig.MAPBOX_NAVIGATION_VERSION_NAME
+        }
+    }
+
+    /**
+     * The lambda that is called if the SDK client did not initialize telemetry. If telemetry is not initialized
+     * than all calls to post a telemetry event will fail with this exception
+     */
+    private val postEventBeforeInit: (TelemetryEventInterface) -> Unit = { _: TelemetryEventInterface ->
+        throw NavigationException("Telemetry must be initialized before calling this method. Call MapboxNavigationTelemetry.initialize()")
+    }
+
+    /**
+     * The lambda that is called once telemetry is initialized. This call will genearate a telemetry event. The call is
+     * equivalent to postEvent()
+     */
+    private val postEventAfterInit: (TelemetryEventInterface) -> Unit = { telemetryInterface -> postEvent(telemetryInterface) }
+
+    /**
+     * The delegate lambda that distaches either a pre or after initialization logic
+     */
+    private var postEventDelegate = postEventBeforeInit
+
+    /**
+     * One-time initializer. Called in responce to initialize() and then replaced with a no-op lambda to prevent multiple initialize() calls
+     */
+    private val primaryInitializer: (Context, String, MapboxNavigation, LocationEngine) -> Unit = { context, token, mapboxNavigation, locationEngine ->
+        this.context = context
+        mapboxToken = token
+        this.locationEngine = locationEngine
+        val options = MapboxNavigationOptions.Builder().build()
+        validateAccessToken(mapboxToken)
+        initializer = postInitialize // prevent primaryInitializer() from being called more than once.
+        postEventDelegate = postEventAfterInit // now that the object has been initialized we can post events
+
+        registerForNotification(mapboxNavigation)
+        monitorChannels()
+        val mapboxTelemetry = MapboxTelemetry(context, token, obtainUserAgent(options))
+        mapboxTelemetry.enable()
+
+        /**
+         * Register a callback to receive location events. At most [MAX_LOCATION_VALUES] are stored
+         */
+        locationEngine.requestLocationUpdates(locationEngineRequest, locationCallback, null)
```

Comment:
Do we want to listen for raw location updates as opposed to enhanced location updates 👀 `LocationObserver`.

Reply:
What was the legacy SDK's behavior? Were we collecting telemetry when navigation was not started? If we weren't we should listen from trip session updates instead.

Reply:
As per our conversation @Guardiola31337, I've removed the code that registers with the LocationEngine to get location change notification and replaced it with registering with MapboxNavigation. The code collects enhanced locations only. If there is a desire collect raw locations, the change is simple